plugins/inputs/cloud_pubsub/pubsub.go

```diff
@@ -175,21 +182,33 @@ func (ps *PubSub) onMessage(ctx context.Context, msg message) error {
 		return fmt.Errorf("message longer than max_message_len (%d > %d)", len(msg.Data()), ps.MaxMessageLen)
 	}
 
+	// This function is called concurrently, but the decoder cannot.
+	ps.decoderMutex.Lock()
+	b, err := ps.decoder.Decode(msg.Data())
+	if err != nil {
+		ps.decoderMutex.Unlock()
+		return fmt.Errorf("unable to decode message: %v", err)
+	}
 	var data []byte
+	if ps.ContentEncoding == "gzip" {
+		data = make([]byte, len(b))
+		copy(data, b)
+	} else {
+		data = b
+	}
+	ps.decoderMutex.Unlock()
 	if ps.Base64Data {
-		strData, err := base64.StdEncoding.DecodeString(string(msg.Data()))
+		strData, err := base64.StdEncoding.DecodeString(string(data))
 		if err != nil {
 			return fmt.Errorf("unable to base64 decode message: %v", err)
 		}
 		data = strData
-	} else {
-		data = msg.Data()
 	}
 
 	metrics, err := ps.parser.Parse(data)
 	if err != nil {
 		msg.Ack()
```

Comment:
I think we cannot ACK messages that fail. You should use `msg.Nack()`.

Reply:
I think this an existing behavior. https://github.com/influxdata/telegraf/blob/master/plugins/inputs/cloud_pubsub/pubsub.go#L191

I think the reason for this behavior is that is if by mistake there is some bad unParsable data enter the pub-sub topic, then we don't want to keep on retrying. Hence we ACK the message and return an error